

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>scirpy.datasets.wu2020 &mdash; scirpy  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/typehints.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="scirpy.datasets.wu2020_3k" href="scirpy.datasets.wu2020_3k.html" />
    <link rel="prev" title="scirpy.pl.styling.style_axes" href="scirpy.pl.styling.style_axes.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/scirpy_logo_bright.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage-principles.html">Usage principles</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="api.html#input-output">Input/Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-scirpy.pp">Preprocessing: <code class="docutils literal notranslate"><span class="pre">pp</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#tools-tl">Tools: <code class="docutils literal notranslate"><span class="pre">tl</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-scirpy.pl">Plotting: <code class="docutils literal notranslate"><span class="pre">pl</span></code></a></li>
<li class="toctree-l2 current"><a class="reference internal" href="api.html#module-scirpy.datasets">Datasets: <code class="docutils literal notranslate"><span class="pre">datasets</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">scirpy.datasets.wu2020</a></li>
<li class="toctree-l3"><a class="reference internal" href="scirpy.datasets.wu2020_3k.html">scirpy.datasets.wu2020_3k</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-scirpy.util">Utility functions: <code class="docutils literal notranslate"><span class="pre">util</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-scirpy.tcr_dist">TCR distance metrics: <code class="docutils literal notranslate"><span class="pre">tcr_dist</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tcr-biology.html">T-cell receptor model</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="zzz_bibliography.html">Bibliography</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">scirpy</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="api.html">API</a> &raquo;</li>
        
      <li>scirpy.datasets.wu2020</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/icbi-lab/scirpy/tree/master/scirpy/datasets/__init__.py#L28-L49" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="scirpy-datasets-wu2020">
<h1>scirpy.datasets.wu2020<a class="headerlink" href="#scirpy-datasets-wu2020" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt id="scirpy.datasets.wu2020">
<code class="sig-prename descclassname">scirpy.datasets.</code><code class="sig-name descname">wu2020</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scirpy.datasets.wu2020" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the dataset from <a class="reference internal" href="zzz_bibliography.html#wu2020" id="id1"><span>[Wu2020]</span></a> as AnnData object.</p>
<p>200k cells, of which 100k have TCRs.</p>
<p>This is how the dataset was processed:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># %load_ext autoreload</span>
<span class="c1"># %autoreload 2</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;../../..&quot;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">scirpy</span> <span class="k">as</span> <span class="nn">st</span>
<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Pool</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">glob</span> <span class="kn">import</span> <span class="n">glob</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="c1"># + language=&quot;bash&quot;</span>
<span class="c1"># mkdir -p data</span>
<span class="c1"># cd data</span>
<span class="c1"># wget -O GSE139555_raw.tar &quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE139555&amp;format=file&quot;</span>
<span class="c1"># wget -O GSE139555_tcell_metadata.txt.gz &quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE139555&amp;format=file&amp;file=GSE139555%5Ftcell%5Fmetadata%2Etxt%2Egz&quot;</span>
<span class="c1"># tar xvf GSE139555_raw.tar</span>

<span class="c1"># + language=&quot;bash&quot;</span>
<span class="c1"># cd data</span>
<span class="c1"># for f in *.matrix.mtx.gz; do</span>
<span class="c1">#   dirname=${f/\.matrix\.mtx\.gz/}</span>
<span class="c1">#   mkdir $dirname</span>
<span class="c1">#   mv $dirname.genes.tsv.gz $dirname/genes.tsv.gz</span>
<span class="c1">#   mv $dirname.matrix.mtx.gz $dirname/matrix.mtx.gz</span>
<span class="c1">#   mv $dirname.barcodes.tsv.gz $dirname/barcodes.tsv.gz</span>
<span class="c1">#   mv $dirname.filtered_contig_annotations.csv.gz $dirname/filtered_contig_annotations.csv.gz</span>
<span class="c1">#   # fix missing feature type column</span>
<span class="c1">#   zcat $dirname/genes.tsv.gz |  awk &#39;{print $0 &quot;\tGene Expression&quot;}&#39; | gzip &gt; $dirname/features.tsv.gz</span>
<span class="c1"># done</span>
<span class="c1"># -</span>

<span class="n">mtx_paths</span> <span class="o">=</span> <span class="n">glob</span><span class="p">(</span><span class="s2">&quot;data/GSM*&quot;</span><span class="p">)</span>

<span class="n">mtx_paths</span>

<span class="n">metadata_all</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="s2">&quot;data/GSE139555_tcell_metadata.txt.gz&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span>
<span class="p">)</span>

<span class="n">umap</span> <span class="o">=</span> <span class="n">metadata_all</span><span class="p">[[</span><span class="s2">&quot;UMAP_1&quot;</span><span class="p">,</span> <span class="s2">&quot;UMAP_2&quot;</span><span class="p">]]</span>

<span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata_all</span><span class="p">[[</span><span class="s2">&quot;ident&quot;</span><span class="p">,</span> <span class="s2">&quot;patient&quot;</span><span class="p">,</span> <span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;source&quot;</span><span class="p">,</span> <span class="s2">&quot;clonotype&quot;</span><span class="p">]]</span>

<span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;clonotype&quot;</span><span class="p">:</span> <span class="s2">&quot;clonotype_orig&quot;</span><span class="p">,</span> <span class="s2">&quot;ident&quot;</span><span class="p">:</span> <span class="s2">&quot;cluster_orig&quot;</span><span class="p">}</span>
<span class="p">)</span>

<span class="n">metadata</span>


<span class="k">def</span> <span class="nf">_load_adata</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="n">sample_id</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="n">obs</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">sample_id</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">umap_coords</span> <span class="o">=</span> <span class="n">umap</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">sample_id</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_mtx</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
    <span class="n">adata_tcr</span> <span class="o">=</span> <span class="n">st</span><span class="o">.</span><span class="n">read_10x_vdj_csv</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;filtered_contig_annotations.csv.gz&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="n">barcode</span><span class="p">)</span> <span class="k">for</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span>
    <span class="p">]</span>
    <span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="n">barcode</span><span class="p">)</span> <span class="k">for</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">adata_tcr</span><span class="o">.</span><span class="n">obs_names</span>
    <span class="p">]</span>
    <span class="c1"># subset to T cells only</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">obs</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">umap_coords</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap_orig&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">umap_coords</span>
    <span class="n">st</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">merge_with_tcr</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adata_tcr</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">adata</span>


<span class="n">p</span> <span class="o">=</span> <span class="n">Pool</span><span class="p">()</span>
<span class="n">adatas</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">_load_adata</span><span class="p">,</span> <span class="n">mtx_paths</span><span class="p">)</span>
<span class="n">p</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">adatas</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">adatas</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>

<span class="c1"># inverse umap X -coordinate</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap_orig&quot;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap_orig&quot;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">])</span> <span class="o">-</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap_orig&quot;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">]</span>
<span class="p">)</span>

<span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;has_tcr&quot;</span><span class="p">])</span>

<span class="n">adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="s2">&quot;wu2020.h5ad&quot;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;lzf&quot;</span><span class="p">)</span>

<span class="n">adata</span><span class="o">.</span><span class="n">obs</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;umap_orig&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cluster_orig&quot;</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s2">&quot;on data&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><span class="annotation terse"><a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.2.dev5+ge4581ab.d20200520)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a></span><span class="annotation full"><a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.2.dev5+ge4581ab.d20200520)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="scirpy.datasets.wu2020_3k.html" class="btn btn-neutral float-right" title="scirpy.datasets.wu2020_3k" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="scirpy.pl.styling.style_axes.html" class="btn btn-neutral float-left" title="scirpy.pl.styling.style_axes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Gregor Sturm, Tamas Szabo.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>