

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Glossary &mdash; scirpy  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/typehints.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bibliography" href="zzz_bibliography.html" />
    <link rel="prev" title="T-cell receptor model" href="tcr-biology.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/scirpy_logo_bright.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage-principles.html">Usage principles</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="tcr-biology.html">T-cell receptor model</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="zzz_bibliography.html">Bibliography</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">scirpy</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Glossary</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/icbi-lab/scirpy/blob/master/docs/glossary.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="glossary">
<h1>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline">¶</a></h1>
<dl class="glossary">
<dt id="term-TCR">TCR</dt><dd><p>T-cell receptor. A TCR consists of one α and one β chain
(or, alternatively, one γ and one δ chain). Each chain consists of a
constant and a variable region. The variable region is responsible for antigen
recognition, mediated by <a class="reference internal" href="#term-CDR"><span class="xref std std-term">CDR</span></a> regions.</p>
<p><em>Scirpy</em> currently only supports α/β-TCRs. For more information, see the
page about our <a class="reference internal" href="tcr-biology.html#tcr-model"><span class="std std-ref">TCR model</span></a>.</p>
<div class="figure align-default" id="id11">
<a class="reference internal image-reference" href="_images/tcr.jpg"><img alt="_images/tcr.jpg" src="_images/tcr.jpg" style="width: 310px;" /></a>
<p class="caption"><span class="caption-text">Image from <a class="reference external" href="https://commons.wikimedia.org/wiki/File:2215_Alpha-Beta_T_Cell_Receptor.jpg">Wikimedia commons</a>
under the <a class="reference external" href="https://creativecommons.org/licenses/by/3.0/deed.en">CC BY-3.0</a> license.</span><a class="headerlink" href="#id11" title="Permalink to this image">¶</a></p>
</div>
</dd>
<dt id="term-Clonotype">Clonotype</dt><dd><p>A clonotype designates a collection of T or B cells that descend from a
common, antecedent cell, and therefore, bear the same adaptive
immune receptors and recognize the same <a class="reference internal" href="#term-Epitope"><span class="xref std std-term">epitopes</span></a>.</p>
<p>In single-cell RNA-sequencing (scRNA-seq) data, T cells sharing identical
complementarity-determining regions 3 (<a class="reference internal" href="#term-CDR3"><span class="xref std std-term">CDR3</span></a>) nucleotide sequences of both α and β
<a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCR</span></a> chains make up a clonotype.</p>
<p><em>Scirpy</em> provides a flexible approach to clonotype definition based on
<a class="reference internal" href="#term-CDR"><span class="xref std std-term">CDR3</span></a> sequence identity or similarity. Additionally, it is possible
to require clonotypes to have the same <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">V-gene</span></a>, enforcing the CDR 1
and 2 regions to be the same.</p>
<p>For more details, see the page about our <a class="reference internal" href="tcr-biology.html#tcr-model"><span class="std std-ref">TCR model</span></a> and
the API documentation of <a class="reference internal" href="generated/scirpy.tl.define_clonotypes.html#scirpy.tl.define_clonotypes" title="scirpy.tl.define_clonotypes"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.define_clonotypes()</span></code></a>.</p>
</dd>
<dt id="term-Clonotype-cluster">Clonotype cluster</dt><dd><p>A higher-order aggregation of <a class="reference internal" href="#term-Clonotype"><span class="xref std std-term">clonotypes</span></a> that have different
<a class="reference internal" href="#term-CDR"><span class="xref std std-term">CDR3</span></a> nucleotide sequences, but might recognize the same antigen
because they have the same or similar CDR3 amino acid sequence.</p>
<p>See also: <a class="reference internal" href="generated/scirpy.tl.define_clonotype_clusters.html#scirpy.tl.define_clonotype_clusters" title="scirpy.tl.define_clonotype_clusters"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.define_clonotype_clusters()</span></code></a>.</p>
</dd>
<dt id="term-Private-clonotype">Private clonotype</dt><dd><p>A clonotype that is specific for a certain patient.</p>
</dd>
<dt id="term-Public-clonotype">Public clonotype</dt><dd><p>A clonotype that is shared across multiple
patients, e.g. a clonotype recognizing common viral epitope.</p>
<div class="figure align-default" id="id12">
<a class="reference internal image-reference" href="_images/public-private.jpg"><img alt="_images/public-private.jpg" src="_images/public-private.jpg" style="width: 420px;" /></a>
<p class="caption"><span class="caption-text">Image from <a class="bibtex reference internal" href="zzz_bibliography.html#setliff2018" id="id1">[SMR+18]</a> under the <a class="reference external" href="https://creativecommons.org/licenses/by/4.0/">CC BY-4.0</a> license.</span><a class="headerlink" href="#id12" title="Permalink to this image">¶</a></p>
</div>
</dd>
<dt id="term-Tissue-specific-clonotype">Tissue-specific clonotype</dt><dd><p>A clonotype that only occurs in a certain tissue of a certain patient.</p>
</dd>
<dt id="term-Multi-tissue-clonotype">Multi-tissue clonotype</dt><dd><p>A clonotype that occurs in multiple tissues of the same patient.</p>
</dd>
<dt id="term-Epitope">Epitope</dt><dd><p>The part of an antigen that is recognized by the <a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCR</span></a>
(or B-cell receptor, or antibody).</p>
</dd>
<dt id="term-CDR3">CDR3</dt><dd><p>Complementary-determining region 3. See <a class="reference internal" href="#term-CDR"><span class="xref std std-term">CDR</span></a>.</p>
</dd>
<dt id="term-CDR">CDR</dt><dd><p>Complementary-determining region. The diversity and, therefore, antigen-specificity
of <a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCRs</span></a> is predominanly determined by three hypervariable loops
(CDR1, CDR2, and CDR3) on each of the α- and β receptor arms.</p>
<p>CDR1 and CDR2 are fully encoded in germline V genes. In contrast,
the CDR3 loops are assembled from V and J segments (TCR-α) and V, D and
J segments (TCR-β) and comprise random additions and deletions at the junction
sites (see also <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">V(D)J</span></a>). Thus, CDR3 regions make up a large part of the
TCR variability and are therefore thought to be particularly important for antigen
specificity (reviewed in <a class="bibtex reference internal" href="zzz_bibliography.html#attaf2015" id="id2">[AHS15]</a>).</p>
<div class="figure align-default" id="id13">
<a class="reference internal image-reference" href="_images/tcr_cdr3.png"><img alt="_images/tcr_cdr3.png" src="_images/tcr_cdr3.png" style="width: 310px;" /></a>
<p class="caption"><span class="caption-text">Image from <a class="bibtex reference internal" href="zzz_bibliography.html#attaf2015" id="id3">[AHS15]</a> under the <a class="reference external" href="https://creativecommons.org/licenses/by-nc-sa/3.0/">CC BY-NC-SA-3.0</a> license.</span><a class="headerlink" href="#id13" title="Permalink to this image">¶</a></p>
</div>
</dd>
<dt id="term-V-D-J">V(D)J</dt><dd><p>The variability of <a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCR</span></a> chain sequences originates from the genetic recombination
of <strong>V</strong>ariable, <strong>D</strong>iversity and <strong>J</strong>oining gene segments. The TCR-α
chain gets assembled from V and J loci only, the TCR-β chain from all three
V, D and J loci.</p>
<p>As an example, the figure below shows how a TCR-α chain is assembed from
the <em>tra</em> locus. V to J recombination joins one of many TRAV segments to one of
many TRAJ segments. Next, introns are spliced out, resulting in a TCR-α chain
transcript with V, J and C segments directly next to each other (reviewed in <a class="bibtex reference internal" href="zzz_bibliography.html#attaf2015" id="id4">[AHS15]</a>).</p>
<div class="figure align-default" id="id14">
<a class="reference internal image-reference" href="_images/vdj.png"><img alt="_images/vdj.png" src="_images/vdj.png" style="width: 600px;" /></a>
<p class="caption"><span class="caption-text">Image from <a class="bibtex reference internal" href="zzz_bibliography.html#attaf2015" id="id5">[AHS15]</a> under the <a class="reference external" href="https://creativecommons.org/licenses/by-nc-sa/3.0/">CC BY-NC-SA-3.0</a> license.</span><a class="headerlink" href="#id14" title="Permalink to this image">¶</a></p>
</div>
</dd>
<dt id="term-Dual-TCR">Dual TCR</dt><dd><p><a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCRs</span></a> with more than one pair of α- and β chains. While this was
previously thought to be impossible due to the mechanism of allelic exclusion
(<a class="bibtex reference internal" href="zzz_bibliography.html#brady2010-gh" id="id6">[BSB10]</a>), there is an increasing amound of evidence for a <em>bona fide</em>
dual-TCR population (<a class="bibtex reference internal" href="zzz_bibliography.html#schuldt2019" id="id7">[SB19]</a>, <a class="bibtex reference internal" href="zzz_bibliography.html#ji2010-bn" id="id8">[JPG10]</a>).</p>
<p>For more information on how <em>Scirpy</em> handles dual TCRs, see the
page about our <a class="reference internal" href="tcr-biology.html#tcr-model"><span class="std std-ref">TCR model</span></a>.</p>
</dd>
<dt id="term-Multichain-cell">Multichain-cell</dt><dd><p>Cells with more than two α- and β chains that do not fit into the
<a class="reference internal" href="#term-Dual-TCR"><span class="xref std std-term">Dual TCR</span></a> model. These are usually rare and could be explained
by doublets/multiplets, i.e. two ore more cells that were captured
in the same droplet.</p>
<div class="figure align-default" id="id15">
<a class="reference internal image-reference" href="_images/multichain.png"><img alt="_images/multichain.png" src="_images/multichain.png" style="width: 450px;" /></a>
<p class="caption"><span class="caption-text">(a) UMAP plot of 96,000 cells from <a class="bibtex reference internal" href="zzz_bibliography.html#wu2020" id="id9">[WMdA+20]</a> with at least one detected
<a class="reference internal" href="#term-CDR3"><span class="xref std std-term">CDR3</span></a> sequence with multichain-cells (n=474) highlighted in green.
(b)  Comparison of detected reads per cell in multichain-cells and other cells.
Multichain cells comprised significantly more reads per cell
(p = 9.45 × 10−251, Wilcoxon-Mann-Whitney-test), supporting the hypothesis that
(most of) multichain cells are technical artifacts arising from
cell-multiplets (<a class="bibtex reference internal" href="zzz_bibliography.html#ilicic2016" id="id10">[IKK+16]</a>).</span><a class="headerlink" href="#id15" title="Permalink to this image">¶</a></p>
</div>
</dd>
<dt id="term-Orphan-chain">Orphan chain</dt><dd><p>A <a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCR</span></a> chain is called <em>orphan</em>, if its corresponding counterpart
has not been detected. For instance, if a cell has only a TCR-α chain,
but no TCR-β chain, the cell will be flagged as “Orphan alpha”.</p>
<p>Orphan chains are most likley the effect of stochastic dropouts due to
sequencing inefficiencies.</p>
<p>See also <a class="reference internal" href="generated/scirpy.tl.chain_pairing.html#scirpy.tl.chain_pairing" title="scirpy.tl.chain_pairing"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.chain_pairing()</span></code></a>.</p>
</dd>
<dt id="term-UMI">UMI</dt><dd><p>Unique molecular identifier. Some single-cell RNA-seq protocols
label each RNA with a unique barcode prior to PCR-amplification to mitigate
PCR bias. With these protocols, UMI-counts replace the read-counts
generally used with RNA-seq.</p>
</dd>
<dt id="term-productive-chain">productive chain</dt><dd><p>Productive chains are TCR chains with a <a class="reference internal" href="#term-CDR3"><span class="xref std std-term">CDR3</span></a> sequence that produces
a functional peptide. Scirpy relies on the preprocessing tools (e.g.
CellRanger or TraCeR) for flagging non-productive chains.
Typically chains are flagged as non-productive if they contain
a stop codon or are not within the reading frame.</p>
</dd>
</dl>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="zzz_bibliography.html" class="btn btn-neutral float-right" title="Bibliography" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="tcr-biology.html" class="btn btn-neutral float-left" title="T-cell receptor model" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Gregor Sturm, Tamas Szabo.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>